source/common/upstream/health_discovery_service.cc

```diff
@@ -68,10 +169,90 @@ void HdsDelegate::onReceiveTrailingMetadata(Http::HeaderMapPtr&& metadata) {
 
 void HdsDelegate::onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) {
   ENVOY_LOG(warn, "gRPC config stream closed: {}, {}", status, message);
-  response_timer_->disableTimer();
+  server_response_timer_->disableTimer();
   stream_ = nullptr;
   handleFailure();
 }
 
+HdsCluster::HdsCluster(Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
+                       const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
+                       Ssl::ContextManager& ssl_context_manager,
+                       Secret::SecretManager& secret_manager, bool added_via_api,
+                       HdsInfoFactory& info_factory)
+    : runtime_(runtime), cluster_(cluster), bind_config_(bind_config), stats_(stats),
+      ssl_context_manager_(ssl_context_manager), secret_manager_(secret_manager),
+      added_via_api_(added_via_api), initial_hosts_(new HostVector()), info_factory_(info_factory) {
+  ENVOY_LOG(debug, "Creating an HdsCluster");
+  priority_set_.getOrCreateHostSet(0);
+
+  ClusterInfoConstSharedPtr info_ptr;
+  info_ptr =
+      info_factory_.createHdsClusterInfo(runtime_, cluster_, bind_config_, stats_,
+                                         ssl_context_manager_, secret_manager_, added_via_api_);
+  info_.swap(info_ptr);
+  for (const auto& host : cluster.hosts()) {
+    initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
+        info_, "", resolveProtoAddress2(host), envoy::api::v2::core::Metadata::default_instance(),
+        1, envoy::api::v2::core::Locality().default_instance(),
+        envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())});
+  }
+  initialize([] {});
+}
+
+ClusterSharedPtr HdsCluster::create() { NOT_IMPLEMENTED; }
+
+void HdsCluster::startHealthChecker(const HealthCheckerSharedPtr& health_checker) {
+  health_checker_ = health_checker;
+  health_checker_->start();
+}
+
+HostVectorConstSharedPtr HdsCluster::createHealthyHostList(const HostVector& hosts) {
+  HostVectorSharedPtr healthy_list(new HostVector());
+  for (const auto& host : hosts) {
+    if (host->healthy()) {
+      healthy_list->emplace_back(host);
+    }
+  }
+  return healthy_list;
+}
+
+ClusterInfoConstSharedPtr RealHdsInfoFactory::createHdsClusterInfo(
+    Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
+    const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
+    Ssl::ContextManager& ssl_context_manager, Secret::SecretManager& secret_manager,
+    bool added_via_api) {
+
+  ClusterInfoConstSharedPtr info;
+  info.reset(new ClusterInfoImpl(cluster, bind_config, runtime, stats, ssl_context_manager,
```

Comment:
This can be written more directly as `return std::make_unique<ClusterInfoImpl>(cluster, ...`